core/primitives/src/utils.rs

```diff
@@ -43,7 +49,105 @@ pub fn get_block_shard_id_rev(
     Ok((block_hash, shard_id))
 }
 
-pub fn create_nonce_with_nonce(base: &CryptoHash, salt: u64) -> CryptoHash {
+/// Creates a new Receipt ID from a given signed transaction and a block hash.
+/// This method is backward compatible, so it takes a current protocol version.
+pub fn create_receipt_id_from_transaction(
+    protocol_version: ProtocolVersion,
+    signed_transaction: &SignedTransaction,
+    block_hash: &CryptoHash,
+) -> CryptoHash {
+    create_hash_upgradable(protocol_version, &signed_transaction.get_hash(), &block_hash, 0)
+}
+
+/// Creates a new Receipt ID from a given receipt, a block hash and a new receipt index.
+/// This method is backward compatible, so it takes a current protocol version.
+pub fn create_receipt_id_from_receipt(
+    protocol_version: ProtocolVersion,
+    receipt: &Receipt,
+    block_hash: &CryptoHash,
+    receipt_index: usize,
+) -> CryptoHash {
+    create_hash_upgradable(protocol_version, &receipt.receipt_id, &block_hash, receipt_index as u64)
+}
+
+/// Creates a new action_hash from a given receipt, a block hash and an action index.
+/// This method is backward compatible, so it takes a current protocol version.
+pub fn create_action_hash(
+    protocol_version: ProtocolVersion,
+    receipt: &Receipt,
+    block_hash: &CryptoHash,
+    action_index: usize,
+) -> CryptoHash {
+    create_hash_upgradable(
+        protocol_version,
+        &receipt.receipt_id,
+        &block_hash,
+        u64::max_value() - action_index as u64,
+    )
+}
+
+/// Creates a new `data_id` from a given action hash, a block hash and a data index.
+/// This method is backward compatible, so it takes a current protocol version.
+pub fn create_data_id(
+    protocol_version: ProtocolVersion,
+    action_hash: &CryptoHash,
+    block_hash: &CryptoHash,
+    data_index: usize,
+) -> CryptoHash {
+    create_hash_upgradable(protocol_version, &action_hash, &block_hash, data_index as u64)
+}
+
+/// Creates a unique random seed to be provided to `VMContext` from a give `action_hash` and
+/// a given `random_seed`.
+/// This method is backward compatible, so it takes a current protocol version.
+pub fn create_random_seed(
+    protocol_version: ProtocolVersion,
+    action_hash: CryptoHash,
```

Comment:
should we pass it as reference here?

Reply:
`clippy` suggest to not pass structs by reference if they implement `Copy`. It's fixed 32 bytes, so it's okay to pass without a reference